### DISCUSSION TITLE
MQTT input to Savvy Can

### DISCUSSION THREAD

**Discussion by nosaj66au** [collin80/SavvyCAN#401] [Q&A]
Hi All, Hoping someone can help me. I'm currently setting up and IOT project and Savvy CAN looks like a useful tool to help along the way with diagnostics. I'm look a using Savvy Can's MQTT feature to monitor MQTT data from my CAN device. I have written a little script to convert the raw data from my USB device and publish to MQTT, however I don't know the Savvy CAN expected payload format. I have used the playback tool/feature and I can see the Savvy Can sending the MQTT packet and it returning them to the live data table, I have also monitored these messages in an MQTT explorer. to this point all has worked as expected. 
However when I attempt to re generate the message and sent it back using an external MQTT publisher, Savvy CAN does not record the incoming message. I have even mimiced the play back message identically. Can anyone help?

**Comment by collin80** [comment]
Just to cover my bases, did you look at pythoncan.py in the root folder of the savvycan sourcecode? That script is what I use to tunnel from socketcan to savvycan over MQTT so if you follow those steps I would have expected it to work.

**Comment by nosaj66au** [comment]
Hi Collin, Thanks for the reply, However I seem to have gone backward and need a little help. I'm using the 204V from about 4 days ago, as I needed unsecured MQTT for testing my application. When I first tested, as per my previous message I was getting MQTT published but not received. Today I have nothing.  :(  I attempt to playback a log file and although its appearing on the capture window nothing is pushed to the MQTT Broker even though i seem to have connection.

![Annotation 2021-07-29 143447](https://user-images.githubusercontent.com/13111932/127432274-f0f54085-9033-40dd-a069-61df5207665b.png)
The broker stats don't show an extra client attaching when opening Savvy Can unlike a Node RED connection I'm testing with. The last connection console message is a worry however as it states MQTT remote Host Closed the Connection? This my be a fingerprint issue? as the IP and port is correct.
  
One other problem, how can I clear the general Savvy Can setting on a different VM, I tried the host name of mqtt:// 10.xx.xx.xx and Savvy Can now crashes before I can remove the "mqtt://" path under preferences, rendering the machine dead for Savvy CAN until I can solve it.
   
I have noticed that the Python script you mentioned in your previous message is not in the SavvyCAN _CIBuild version (204)?

Thanks again for your help.
Jason

**Comment by nosaj66au** [comment]
Hi Collin,
   Fixed the problems all round, You need a subscription and publish topic for it to work. However I'm still getting the Closed Connection message which is a but misleading.
Thanks for you help..
Jason
